File: programmers/게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque


def solution(maps):
    n = len(maps)
    m = len(maps[0])

    y_value = [-1, 0, 1, 0]
    x_value = [0, 1, 0, -1]

    deq = deque()
    # y, x, 거리
    deq.append([0, 0, 1])

    visited = [[False] * m for _ in range(n)]
    visited[0][0] = True

    while deq:
        start = deq.popleft()

        for i in range(4):
            start_y, start_x = start[0] + y_value[i], start[1] + x_value[i]

            if 0 > start_y or start_y >= n or 0 > start_x or start_x >= m:
                continue

            if not maps[start_y][start_x]:
                continue

            if visited[start_y][start_x]:
                continue

            if start_y == n - 1 and start_x == m - 1:
                return start[2] + 1

            visited[start_y][start_x] = True
            deq.append([start_y, start_x, start[2] + 1])

    return -1


print(solution([
    [1, 0, 1, 1, 1],
    [1, 0, 1, 0, 1],
    [1, 0, 1, 1, 1],
    [1, 1, 1, 0, 1],
    [0, 0, 0, 0, 1]],
))

print(solution([
    [1, 0, 1, 1, 1],
    [1, 0, 1, 0, 1],
    [1, 0, 1, 1, 1],
    [1, 1, 1, 0, 0],
    [0, 0, 0, 0, 1]],
))

File: programmers/게임 맵 최단거리/게임 맵 최단거리.java
import java.util.ArrayDeque;

public class Main {
    public static void main(String[] args){
        Solution solution = new Solution();
        int[][] maps = {{1,0,1,1,1}, {1,0,1,0,1}, {1,0,1,1,1}, {1,1,1,0,1}, {0,0,0,0,1}};
        System.out.println(solution.solution(maps));
    }
}

class Solution {
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        return bfs(maps, n, m);
    }

    private static int bfs(int[][] maps, int n, int m) {
        boolean[][] visited = new boolean[n][m];
        visited[0][0] = true;

        int[] yValue = {-1, 0, 1, 0};
        int[] xValue = {0, 1, 0, -1};

        ArrayDeque<int[]> deq = new ArrayDeque<>();
        deq.add(new int[]{0, 0, 1});

        while (!deq.isEmpty()) {
            int[] item = deq.pollFirst();
            int y = item[0];
            int x = item[1];
            int cnt = item[2];

            for (int i = 0; i < 4; i++) {
                int yIdx = y + yValue[i];
                int xIdx = x + xValue[i];

                if (0 > yIdx || yIdx >= n || 0 > xIdx || xIdx >= m) {
                    continue;
                }

                if (yIdx == n - 1 && xIdx == m - 1) {
                    return cnt + 1;
                }

                if (visited[yIdx][xIdx]) {
                    continue;
                }

                if (maps[yIdx][xIdx] == 0) {
                    continue;
                }

                visited[yIdx][xIdx] = true;
                deq.add(new int[]{yIdx, xIdx, cnt + 1});
            }
        }

        return -1;
    }
}
